src/core/blocks/serialization.cpp

```diff
@@ -385,9 +387,75 @@ struct WritePNG : public FileBase {
     int w = int(input.payload.imageValue.width);
     int h = int(input.payload.imageValue.height);
     int c = int(input.payload.imageValue.channels);
-    if (0 == stbi_write_png(filename.c_str(), w, h, c,
-                            input.payload.imageValue.data, w * c))
-      throw ActivationError("Failed to write PNG file.");
+
+    // demultiply alpha if needed, limited to 4 channels
+    if (c == 4 &&
+        (input.payload.imageValue.flags & CBIMAGE_FLAGS_PREMULTIPLIED_ALPHA) ==
+            CBIMAGE_FLAGS_PREMULTIPLIED_ALPHA) {
+      _scratch.resize(w * h * 4 * pixsize);
+      int lineSize = w * 4;
+      for (int ih = 0; ih < h; ih++) {
+        for (int iw = 0; iw < w; iw++) {
+          if (pixsize == 1) {
+            uint8_t a =
+                input.payload.imageValue.data[(ih * lineSize) + (iw * 4) + 3];
+            float fa = float(a) / 255.0f;
+            _scratch[(ih * lineSize) + (iw * 4) + 3] = a;
+            uint8_t r =
+                input.payload.imageValue.data[(ih * lineSize) + (iw * 4) + 0];
+            _scratch[(ih * lineSize) + (iw * 4) + 0] =
+                uint8_t(float(r) / fa + 0.5);
```

Comment:
Might save a few CPU cycles to do the index calc only once:
```cpp
auto index = (ih * lineSize) + (iw * 4) + 0;
uint8_t r = input.payload.imageValue.data[index];
 _scratch[index] = uint8_t(float(r) / fa + 0.5);
```

Reply:
I think any half decent compiler will optimize those really but for readability let's do that